src/software/ai/hl/stp/tactic/goalie_tactic_test.cpp

```diff
@@ -71,3 +77,122 @@ TEST_P(GoalieRestrainTest, goalie_position_safe)
+        auto action_ptr = tactic.getNextAction();
+
+        EXPECT_TRUE(action_ptr);
+
+        auto stop_action = std::dynamic_pointer_cast<StopAction>(action_ptr);
+        ASSERT_NE(stop_action, nullptr);
+    }
+    void expectChipAction(Ball ball)
+    {
+        World world         = ::Test::TestUtil::createBlankTestingWorld();
+        world.mutableBall() = std::move(ball);
+
+        Robot goalie = Robot(0, Point(-4.5, 0), Vector(0, 0), Angle::zero(),
+                             AngularVelocity::zero(), Timestamp::fromSeconds(0));
+        world.mutableFriendlyTeam().updateRobots({goalie});
+        world.mutableFriendlyTeam().assignGoalie(0);
+
+        GoalieTactic tactic = GoalieTactic(world.ball(), world.field(),
+                                           world.friendlyTeam(), world.enemyTeam());
+        tactic.updateRobot(goalie);
+        auto action_ptr = tactic.getNextAction();
+
+        EXPECT_TRUE(action_ptr);
+
+        auto chip_action = std::dynamic_pointer_cast<ChipAction>(action_ptr);
+        ASSERT_NE(chip_action, nullptr);
+        EXPECT_TRUE(chip_action->getChipOrigin().isClose(world.ball().position(), 0.001));
+        EXPECT_EQ(chip_action->getChipDirection(),
+                  (world.ball().position() - world.field().friendlyGoal()).orientation());
+        EXPECT_NEAR(chip_action->getChipDistanceMeters(), 2, 0.001);
+    }
+};
+
+TEST_F(GoalieTacticTest, ball_very_fast_in_straight_line)
+{
+    Ball ball = Ball(Point(0, 0), Vector(-4, 0), Timestamp::fromSeconds(0));
+    expectMoveAction(ball, Point(-4.5, 0));
+}
+
+TEST_F(GoalieTacticTest, ball_very_fast_in_diagonal_line)
+{
+    Ball ball = Ball(Point(0, 0), Vector(-4.5, 0.25), Timestamp::fromSeconds(0));
+    expectMoveAction(ball, Point(-4.5, 0.25));
+}
+
+TEST_F(GoalieTacticTest, ball_very_fast_miss)
+{
+    Ball ball = Ball(Point(0, 0), Vector(-4.5, 1), Timestamp::fromSeconds(0));
+    // Goalie is expected to default to centre of goal
+    expectMoveAction(ball, Point(-3.7, 0));
+}
+
+TEST_F(GoalieTacticTest, ball_slow_inside_dont_chip_rectangle)
+{
+    Ball ball = Ball(Point(-4.5 + ROBOT_MAX_RADIUS_METERS, 0), Vector(-0.1, 0.1),
+                     Timestamp::fromSeconds(0));
+    expectStopAction(ball);
+}
+
+TEST_F(GoalieTacticTest, ball_slow_outside_dont_chip_rectangle)
+{
+    Ball ball = Ball(Point(-3.5, 0.5), Vector(-0.1, -0.1), Timestamp::fromSeconds(0));
+
+    expectChipAction(ball);
+}
+
+TEST_F(GoalieTacticTest,
+       ball_behind_net_and_moving_toward_net)  // snap to closer goal post
+{
+    Ball ball = Ball(Point(-5.8, 0.3), Vector(0.5, 0.5), Timestamp::fromSeconds(0));
+    expectMoveAction(ball, Point(-4.5, 0.5 - ROBOT_MAX_RADIUS_METERS));
+}
+
+TEST_F(GoalieTacticTest,
+       ball_angle_very_sharp_and_low_velocity)  // snap to closer goal post
+{
+    Ball ball = Ball(Point(-4.5, -3), Vector(0, 0.1), Timestamp::fromSeconds(0));
+    expectMoveAction(ball, Point(-4.5, -0.5 + ROBOT_MAX_RADIUS_METERS));
+}
```

Comment:
nit: missing newline at end of file

Reply:
fixed but just curious is the formatting script unable to find this error?

Reply:
No because it's not strictly a formatting error in that sense. Here's a better explanation of why it matters (mostly for git / github's sake)

https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

Reply:
No because it's not strictly a formatting error in that sense. Here's a better explanation of why it matters (mostly for git / github's sake)

https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file